Propose  BMH specific preprovisioningExtraKernelParams spec field (#552)
metal3-io/metal3-docs: design/baremetal-operator/node-specific-kernel-args.md

<!--
 This work is licensed under a Creative Commons Attribution 3.0
 Unported License.

 http://creativecommons.org/licenses/by/3.0/legalcode
-->

# pre-provisioning kernel argument support

## Status

planned

## Summary

This document proposes a new BareMetalHost spec attribute intended for
configuring the pre-provisioning image's (usually IPA) kernel arguments.

## Motivation

There are situations where a user would need node specific kernel arguments
for the pre-provisioning image. The reason could be the use of a mixed set of
hardware, troubleshooting provisioning or boot issues, selectively configuring
custom logic e.g. enabling IPA plugins or other services embedded in the
pre-provisioning image.

### Goals

1. Implement a string array type optional spec field named

>    `preProvisionKernelArgs`

Reviewer: Please keep spelling and word choice consistent: `preprovisioningKernelParams`

Reviewer: Actually, it's called `ExtraKernelParams` in PPI. There is a reason for that: these options are appended to the defaults in ironic-image. I'd prefer to keep it this way for the new feature, otherwise it won't be usable for anyone with ironic-image-level customizations (like we have to provide something for CoreOS/Ignition).

2. Extend the generic `Provisioner` interface to support the new optional field
3. Implement support for the `preProvisionKernelArgs` in the Ironic provisioner

> 

Reviewer: You also need to extend the PreprovisioningImage object to accept these params.

Reviewer: That will create a wild inconsistency that will cause users file bugs for me with `preprovisioningExtraKernelParams` does not work :(

Reviewer: Btw it's fair if you ask me to follow-up with PPI support myself if you don't have cycles for that. I just want it to be in the design.

### Non-Goals

> 1. To implement any sort value check for the new optional spec

Reviewer 2: I guess *some* kind of validation may be done but I take it this means that we are not going to check the content, only that it is a string array?
```suggestion
1. To implement any sort validation for the new optional field
```

2. Providing a default value for the new spec field

## Proposal

This document proposes a new BareMetalHost spec attribute intended for
configuring the pre-provisioning image's (usually IPA) kernel arguments.

The proposal involves the introduction of a new spec field thus a modification
to the BMO API. The API change is intended to be backward compatible as the

> new `preProvisionKernelArgs` field intended to be optional.

Reviewer 2: ```suggestion
new `preProvisionKernelArgs` field is optional.
```

### Implementation Details/Notes/Constraints

The `preProvisionKernelArgs` field would be located directly under `spec`
section. The format of the new field would be a list/array of individual
strings. The new field would be optional and it would be considered
either a NOOP or a cleaning operation depending on the previous value of the
field in case it would be not present or have no value or would have an empty
array as a value.

The proposed BMH looks like the following:

```yaml
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
 name: bm0
spec:
 online: true
 bmc:
   address: idrac://192.168.122.1:6230/
   credentialsName: bm0-bmc-secret

>  preProvisionKernelArgs:

Author: I have changed it to a string.

   - --timeout
   - 60000
   - systemd.journald.forward_to_console=yes
   - ipa-debug=1
   ...
   ...
   ...
 bootMACAddress: 52:54:00:b7:b2:6f
...
...
...

```

The example contains a snippet of a common IPA kernel argument list.
The `preProvisionKernelArgs` list would be received by the `Provisioner`
interface as is. The handling of the list would be up to the particular
implementations of the `Provisioner` interface. In case of the `Ironic`
provisioner the elements from the `preProvisionKernelArgs` would be

> concatenated into a single string such that the string would start with the

Reviewer 2: They are concatenated with spaces in between, right? I think that may be worth to point out explicitly here also.

first element (lowest index) from the list and would end with the last
(highest index) element. The elements between the first and last element would
be concatenated in an ascending order based on their index number in the array.

> Taken the BMH snippet from above the kernel argument list generated by

Reviewer 2: ```suggestion
Given the BMH snippet from above, the kernel argument list generated by
```

Given the BMH snippet from above, the kernel argument list generated by
the `Ironic provisioner` and posted to the Ironic Node API  would look like
`--timeout 60000 systemd.journald.forward_to_console=yes ipa-debug=1 ...`.

### Risks and Mitigations

Providing a faulty list kernel arguments might result in boot issues
or unexpected behavior during the life cycle of the pre-provisioning image.

The user has to be aware that the kernel arguments supplied here will fully or
partially overwrite the default arguments depending on how the particular
`Provisioner` is handling the argument list. In case of Ironic provisioner
the `preProvisionKernelArgs` will completely replace the default argument

> list used by Ironic.

Reviewer: See above, I'd prefer to avoid a full override. It will require a user to guess Ironic options that must stay present.

Author: I have dropped the full override approach for this proposal.

## Design Details

> Most of the design is very straightforward an easy to combine with the

Reviewer 2: ```suggestion
Most of the design is very straightforward and easy to combine with the
```

reconcile logic, the support for node specific pre-provisioning kernel args
also already exist in Ironic, so right after the feature is implemented in BMO
it can be used.

The only challenge is the process of updating/cleaning the previously applied
kernel argument list, this cleanup/update process highly depends on the
particular provisioner but we might be able to leverage the
`kubectl.kubernetes.io/last-applied-configuration` annotation or some other
method to avoid initiating redundant kernel argument update,sync or cleaning
operations.

### Work Items

- Implement a string array type optional spec field named
  `preProvisionKernelArgs`
- Extend the generic `Provisioner` interface to support the new optional field
- Implement support for the `preProvisionKernelArgs` in the Ironic driver
- Implement unit tests
- Add a check to the existing BMO e2e workflow that logs in to an active IPA
  via ssh or serial console and checks whether the custom kernel args were

>   propagated properly or not.

Reviewer 2: :+1: 
I have recently learned also that it is possible to use the kernel command line to [configure the network](https://www.kernel.org/doc/Documentation/filesystems/nfs/nfsroot.txt). Maybe we could use that to verify also, by configuring a static IP outside of DHCP range? If IPA becomes reachable at that IP we can consider the test passed.

### Dependencies

- Ironic in case of the `Ironic Provisioner`

### Test Plan

- Unit tests for the functions

> 

Reviewer 2: ```suggestion
```

<!--
 This work is licensed under a Creative Commons Attribution 3.0
 Unported License.

 http://creativecommons.org/licenses/by/3.0/legalcode
-->

# pre-provisioning kernel argument support

## Status

planned

## Summary

This document proposes a new BareMetalHost spec attribute intended for
configuring the pre-provisioning image's (usually IPA) kernel arguments.

## Motivation

There are situations where a user would need node specific kernel arguments
for the pre-provisioning image. The reason could be the use of a mixed set of
hardware, troubleshooting provisioning or boot issues, selectively configuring
custom logic e.g. enabling IPA plugins or other services embedded in the
pre-provisioning image.

### Goals

1. Implement a string type optional spec field named `preprovisioningExtraKernelParams`
2. Extend the generic `Provisioner` interface to support the new optional field
3. Implement support for the `preprovisioningExtraKernelParams` in the Ironic provisioner
4. Combine the values of the `preprovisioningExtraKernelParams` and the
   `preprovisioningImage`'s `status.extraKernelParams` field if a

>    `preprovisioningImage` is refferences by the BMH and teh `status.extraKernelParams`

Reviewer 2: ```suggestion
   `preprovisioningImage` is referenced by the BMH and the `status.extraKernelParams`
```

>    is not empty

Reviewer 3: Is it a good idea to construct a value of a field from two different sources? I imagine things might go wrong, for example how do we handle a case when prepovisioningImage object referenced by BMH doesn't exist for some reason?

Does this mean that a node might have different kernel parameters compared to what is asked it via `.spec.preprovisioningExtraKernelParams`, in case  preprovisioningImage `status.extraKernelParams` has something non-empty? Should not we have a single source of truth for defining the parameters and I guess the status sub-resource was never meant for defining the parameters but more for summarizing the current state right? Perhaps whoever writes into `status.extraKernelParams` should only rely on BMH spec and not something else? Or this a transitional state that we eventually target to move to a single source?

5. Introducing a `preprovisioningExtraKernelParams` related status field that
   would display the last applied `preprovisioningExtraKernelParams` value.

>    Let's call this field `status.lastAppliedPreprovisioningExtraKernelParams`

Reviewer 3: Perhaps something shorter would be easier but I don't have strong opinion on this
Example, `status.lastAppliedKernelParams`

### Non-Goals

1. To implement any sort validation for the new optional field
2. Providing a default value for the new spec field
3. Implement conditional logic to overwrite default kernel parameters on
   per node basis

## Proposal

This document proposes a new BareMetalHost spec attribute intended for
configuring the pre-provisioning image's (usually IPA) kernel arguments.

The proposal involves the introduction of a new spec field thus a modification
to the BMO API. The API change is intended to be backward compatible as the
new `preprovisioningExtraKernelParams` field is optional.

### Implementation Details/Notes/Constraints

The `preprovisioningExtraKernelParams` field would be located directly under `spec`

> section. The format of the new field would be a list/array of individual

Reviewer 2: This has been changes elsewhere to be a string

```suggestion
section. The format of the new field would be a string.
The new field would be optional and it would be considered
```

strings. The new field would be optional and it would be considered
either a NOOP or a cleaning operation depending on the previous value of the
field in case it would be not present or have no value or would have an empty
array as a value.

The proposed BMH looks like the following:

```yaml
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
 name: bm0
spec:
 online: true
 bmc:
   address: idrac://192.168.122.1:6230/
   credentialsName: bm0-bmc-secret

>  preprovisioningExtraKernelParams: "--timeout 60000 systemd.journald.forward_to_console=yes ipa-debug=1"

Reviewer 2: ```suggestion
  preprovisioningExtraKernelParams: "--timeout 60000 systemd.journald.forward_to_console=yes ipa-debug=1"
```

| No PPI / No PPI Support    | Used        | N/A        | `%default% BMH_params` | Standard behavior |
| ISO (External)             | Ignored     | Used       | `PPI_params` | Warning logged if BMH params specified |
| ISO (BMO-managed)          | Used        | Ignored    | `%default% BMH_params` | Consistent with no-PPI behavior |
| InitRD (External)          | Used        | Used       | `%default% BMH_params PPI_params` | Parameters combined |
| InitRD (BMO-managed)       | Used        | Ignored    | `%default% BMH_params` | Consistent with no-PPI behavior |
| Unknown Format             | Used        | Ignored    | `%default% BMH_params` | Fallback behavior |

### Risks and Mitigations

Providing faulty kernel arguments might result in boot issues or unexpected
behavior during the lifecycle of the preprovisioning image. The feature
provides several mitigations:

- Graceful fallback to BMH parameters when PPI processing fails
- Warning logs for format-specific parameter behavior
- Maintains %default% prefix for Ironic consistency
- No validation prevents breaking existing workflows

## Design Details

> The feature extends the BMO architecture with:

Reviewer: nit: this is too detailed for a design proposal IMO
